### DISCUSSION TITLE
ImGui and MSAA

### DISCUSSION THREAD

**Discussion by rolandhill** [vsg-dev/VulkanSceneGraph#758] [Q&A] [answered]
Hi All,

I was wondering if anybody has got ImGui to work with MSAA? I have ImGui working well in my application already. When I switch on MSAA using

    windowTraits->samples = VK_SAMPLE_COUNT_8_BIT;
    windowTraits->vulkanVersion = VK_API_VERSION_1_2;

my general 3D scene looks great, but the ImGui windows are squashed up into a corner and I get a heap of validation errors about VkPipelineMultisampleStateCreateInfo::rasterizationSamples must be the same as the current subpass.

I know it's because ImGui renders straight onto the screen space and I need to resolve the multisample frame buffer before passing it to ImGui, I'm just not sure how to go about it. Has anyone else tackled this successfully?

Thanks,
Roland

**Comment by timoore** [comment]
Good question. https://github.com/ocornut/imgui/pull/2706 addresses this question from the ImGui side. After a very quick look, I think you should be able to change the setup code in RenderImGui.cpp (_init()) to pass a multisample count to ImGui.

**Comment by vsg-dev** [comment]
A quick search of vsgImGui comes up with the following code in imgui_impl_vulkan,cpp:


    VkPipelineMultisampleStateCreateInfo ms_info = {};
    ms_info.sType = VK_STRUCTURE_TYPE_PIPELINE_MULTISAMPLE_STATE_CREATE_INFO;
    ms_info.rasterizationSamples = (MSAASamples != 0) ? MSAASamples : VK_SAMPLE_COUNT_1_BIT;

**Comment by vsg-dev** [comment]
I decided to use this issue to finish up some work I did on vsgImGui last month but hadn't merge with master, and to update imgui to the latest v1.89.4 release and implot to master to keep things compiling. This work is in the ImGuiNode branch:

https://github.com/vsg-dev/vsgImGui/tree/ImGuiNode

I have also checked into vsgExamples master the addition of --samples <num_samples> to the vsgimgui_example to test the issue that @rolandhill reported and the issue appears with the ImGui dialog appearing tiny :-)

I will now look into fixes the issue.

**Comment by vsg-dev** [comment, marked as answer]
I have got the MSAA working with the commit to the ImGuiNode branch:

https://github.com/vsg-dev/vsgImGui/commit/14c9dc874753cff32275f0c27db58a1f4b632bc1

I need to reflect on this branch before merging with vsgImGui master.

**Comment by vsg-dev** [comment]
I have adapted @timoore work on adding texture images to vsgImGui to work with my refactor that I did in the ImGuiNode branch, and have adapted Tim's changes to vsgimgui_example to work with this new scheme as well.  Changes can be found in the ImGuiNode branch of vsgImGui and vsgExamples:

https://github.com/vsg-dev/vsgImGui/tree/ImGuiNode
https://github.com/vsg-dev/vsgExamples/tree/ImGuiNode

`$ vsgimgui_example models/openstreetmap.vsgt
`
![Screenshot_20230321_140750](https://user-images.githubusercontent.com/44898258/226631985-278fe81e-ee7b-4a75-ae48-4340b296be4a.png)

Could folks test out the above ImGuiNode branches as this should make vsgImGui more flexible, support MSAA and texture images.  if that works fine then I'll merge with vsgImGui and vsgExamples master.

**Reply by robertosfield** [reply]
I have continued to refine the ImGuiNode branch of vsgImGui and vsgExamples, I'm trying to simplify usage and implementation while make it more flexible.  I've run out of day to wrap this work up so will return to it tomorrow.  Testing out these two branches is still helpful as an feedback/insights that you can provide I can help shape the final elements of work.

**Comment by rolandhill** [comment]
Hi Robert,

I tested the ImGuiNode branches of vsgImGui and vsgExamples. The
vsgimgui_example example worked perfectly. I then tested my own application
and it works perfectly too. I don't see how it could get any simpler - just
set windowTraits->samples to the value you want and it's done. I notice
that setting windowTraits->vulkanVersion = VK_API_VERSION_1_2 is not
required anymore either.

The only other thing I noticed is that I get errors building vsgExamples
for vsgviewer and vsgmeshshader. They also threw errors on the master
branch.

Thanks for the fantastic support!

Regards,
Roland

**Reply by robertosfield** [reply]
That's good to hear.

The compile error with vsgviewer and vsgmeshshader suggest to me that you need to update your VSG version to 1.0.4 or master.

**Comment by rolandhill** [comment]
Thanks for the response, Tim. I did search for this on the ImGui pages but
didn't come up with that issue, only older ones suggesting I had to resolve
the multisamples first.

As you will have seen, Robert has addressed this now in the ImGuiNode
branch. I've tested it and it's working well now.

Cheers,
Roland

**Comment by rolandhill** [comment]
I synced my fork with VSG then pulled from my fork. I'm using the master
branch. Git Graph shows the following
[image: image.png]

Maybe V1.0.4 has moved ahead of master on the VSG repo.

Cheers,
Roland

**Reply by robertosfield** [reply]
I have just done a diff and VulkanSceneGraph-1.0.4 and VSG master are the same.

**Comment by robertosfield** [comment]
I have completed my refactor of vsgImGui to make it more flexible, update to ImGui v 1.89.4 and ImPlot master and to support image texture using a new vsgImGui::Texture class:

[Changes to vsgImGui](https://github.com/vsg-dev/vsgImGui/pull/47)
[Changes to vsgExamples](https://github.com/vsg-dev/vsgExamples/pull/218)

This also has the support for handling the MSAA.  I have updated the vsgImGui version to 0.1.0 and vsgExamples to 1.0.4 to reflect these changes.  I will post more info on a separate Discussion thread for the vsgImGui thread as this thread really about the MSAA issues.

**Comment by rolandhill** [comment]
Thanks, Robert. I just repulled and rebuilt everything using master
branches and it all works. I did it before, but must have screwed up
somewhere along the way.

Cheers,
Roland
